lib/northbound.c

```diff
@@ -1664,79 +1654,18 @@ int nb_notification_send(const char *xpath, struct list *arguments)
 	return ret;
 }
 
-/* Running configuration user pointers management. */
-struct nb_config_entry {
-	char xpath[XPATH_MAXLEN];
-	void *entry;
-};
-
-static bool running_config_entry_cmp(const void *value1, const void *value2)
-{
-	const struct nb_config_entry *c1 = value1;
-	const struct nb_config_entry *c2 = value2;
-
-	return strmatch(c1->xpath, c2->xpath);
-}
-
-static unsigned int running_config_entry_key_make(const void *value)
-{
-	return string_hash_make(value);
-}
-
-static void *running_config_entry_alloc(void *p)
-{
-	struct nb_config_entry *new, *key = p;
-
-	new = XCALLOC(MTYPE_NB_CONFIG_ENTRY, sizeof(*new));
-	strlcpy(new->xpath, key->xpath, sizeof(new->xpath));
-
-	return new;
-}
-
-static void running_config_entry_free(void *arg)
-{
-	XFREE(MTYPE_NB_CONFIG_ENTRY, arg);
-}
-
 void nb_running_set_entry(const struct lyd_node *dnode, void *entry)
 {
-	struct nb_config_entry *config, s;
-
-	yang_dnode_get_path(dnode, s.xpath, sizeof(s.xpath));
-	config = hash_get(running_config_entries, &s,
-			  running_config_entry_alloc);
-	config->entry = entry;
-}
-
-static void *nb_running_unset_entry_helper(const struct lyd_node *dnode)
-{
-	struct nb_config_entry *config, s;
-	struct lyd_node *child;
-	void *entry = NULL;
-
-	yang_dnode_get_path(dnode, s.xpath, sizeof(s.xpath));
-	config = hash_release(running_config_entries, &s);
-	if (config) {
-		entry = config->entry;
-		running_config_entry_free(config);
-	}
-
-	/* Unset user pointers from the child nodes. */
-	if (CHECK_FLAG(dnode->schema->nodetype, LYS_LIST | LYS_CONTAINER)) {
-		LY_TREE_FOR (dnode->child, child) {
-			(void)nb_running_unset_entry_helper(child);
-		}
-	}
-
-	return entry;
+	lyd_set_private(dnode, entry);
```

Comment:
In the previous commit hash table was introduced and private pointer was removed as priv pointer might be lost and a function nb_candidate_restore_priv_pointers would get it back but was expensive. By not doing nb_candidate_restore_priv_pointers will we not have private pointer lost?

Reply:
lyd_set_private returns previous set pointer if any. I think it makes sense to return that to use in case if that needs to be freed.

Reply:
> In the previous commit hash table was introduced and private pointer was removed as priv pointer might be lost and a function nb_candidate_restore_priv_pointers would get it back but was expensive. By not doing nb_candidate_restore_priv_pointers will we not have private pointer lost?

No because the new `yang_dnode_replace()` function won't delete dnodes that don't need to be deleted, thus preserving their user pointers.

> lyd_set_private returns previous set pointer if any. I think it makes sense to return that to use in case if that needs to be freed.

That should never happen in practice because `nb_running_set_entry()` is supposed to be called by `create()` callbacks only, in which case `dnode->priv` will always be null.